qiskit/assembler/assemble_schedules.py

```diff
@@ -53,21 +55,26 @@ def assemble_schedules(schedules, qobj_id, qobj_header, run_config):
 
     # Pack everything into the Qobj
     qobj_schedules = []
-    user_pulselib = set()
+    user_pulselib = {}
     for idx, schedule in enumerate(schedules):
         # instructions
         qobj_instructions = []
         # Instructions are returned as tuple of shifted time and instruction
         for shift, instruction in schedule.instructions:
             # TODO: support conditional gate
-            qobj_instructions.append(instruction_converter(shift, instruction))
             if isinstance(instruction, PulseInstruction):
+                name = instruction.command._name
+                if name in user_pulselib and not np.allclose(instruction.command.samples,
+                                                             user_pulselib[name]):
+                    name = "{0}-{1:x}".format(name, hash(instruction.command.samples.tostring()))
                 # add samples to pulse library
-                user_pulselib.add(instruction.command)
+                user_pulselib[name] = instruction.command.samples
+                instruction.command._name = name
```

Comment:
I think it would be better to create a new SamplePulse here with a new name and the same samples, rather than mutate the existing SamplePulse.

Reply:
Are we guaranteed to always have a Sample Pulse command and no other kinds of instructions?

Reply:
(as the command within a PulseInstruction)

Reply:
Running into another issue: I can't set `instruction.command` either, so I tried to make a new `PulseInstruction`, which complained because `timeslots` isn't an initarg... but I'm guessing it's important to carry that information forward.  I will look further into this tomorrow, but lmk what you'd do if you know offhand.

Reply:
Also, `instruction = ...` should give `qobj_instructions.append(instruction_converter(shift, instruction))` the right behavior, but do I have to worry about the fact that the original schedule isn't going to have the correct name for that `PulseInstruction`?

Reply:
I think you're on the right track, will need to create a new `PulseInstruction` and `SamplePulse`  due to immutability. Should be something like 
```PulseInstruction(SamplePulse(old_samples, name=new_name), old_instruction.channels[0])```